applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
+        self.adjoint_analysis._GetSolver().SolveSolutionStep()
+        self.adjoint_analysis.FinalizeSolutionStep()
+        self.adjoint_analysis.OutputSolutionStep()
+        print("> Time needed for solving the adjoint analysis = ",round(timer.time() - startTime,2),"s")
+
+
+    def GetValue(self):
+        return self.primal_model_part.ProcessInfo[StructuralMechanicsApplication.RESPONSE_VALUE]
+
+
+    def GetShapeGradient(self):
+        gradient = {}
+        for node in self.adjoint_model_part.Nodes:
+            gradient[node.Id] = node.GetSolutionStepValue(SHAPE_SENSITIVITY)
+        return gradient
+
+
+    def FinalizeSolutionStep(self):
+        # TODO reset DISPLACEMENT, ROTATION ADJOINT_DISPLACEMENT and ADJOINT_ROTATION
+        pass
+
+
+    def Finalize(self):
+        self.primal_analysis.Finalize()
+        self.adjoint_analysis.Finalize()
+
+# ==============================================================================
+class AdjointStrainEnergyResponse(ResponseFunctionBase):
+    """Linear static adjoint response function.
+    - runs the primal analysis (writes the primal results to an .h5 file)
+    - reads the primal results from the .h5 file into the adjoint model part
+    - uses primal results to calculate value
+    - uses primal results to calculate gradient
+
+    Note: because of the special property of the adjoint strain energy response,
+    no adjoint solution is necessary. except from that it works the same as
+    'AdjointResponseFunction'.
+
+    Attributes
+    ----------
+    primal_analysis : Primal analysis object of the response function
+    response_function_utility: Cpp utilities object doing the actual computation of response value and gradient.
+    """
+    def __init__(self, identifier, project_parameters, model_part = None):
+        self.identifier = identifier
+        self.response_function_utility = StructuralMechanicsApplication.AdjointStrainEnergyResponseFunction( model_part, project_parameters )
+
+        model = Model()
+        model.AddModelPart(model_part)
+        self.primal_model_part = model_part
+
+        # Create the primal solver
+        self.ProjectParametersPrimal = Parameters( open(project_parameters["primal_settings"].GetString(),'r').read() )
+        self.primal_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(model, self.ProjectParametersPrimal)
+
+        self.primal_model_part.AddNodalSolutionStepVariable(SHAPE_SENSITIVITY)
+
+        # Add variables to save the solution of the adjoint problem
+        self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT)
+        if self.ProjectParametersPrimal["solver_settings"]["rotation_dofs"].GetBool():
+            self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.ADJOINT_ROTATION)
+        #self.primal_model_part.AddNodalSolutionStepVariable(StructuralMechanicsApplication.POINT_LOAD_SENSITIVITY)
```

Comment:
This is a good question, this will give segfaults in the worst case!

Reply:
Yes needs to be added.